src/App.tsx

```diff
@@ -1,24 +1,212 @@
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch((fetchedError: Error) => {
+        setError(fetchedError?.message ?? 'Unexpected error fetching todos');
+      });
+  }, []);
+
+  useEffect(() => {
+    let timeout: number;
+
+    if (error) {
+      timeout = window.setTimeout(() => setError(null), 3000);
+    }
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [error]);
+
+  const activeTodos = useMemo(() => (
+    todos.filter(todo => !todo.completed)
+  ), [todos]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter(todo => todo.completed)
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return getVisibleTodos(filter, todos, completedTodos, activeTodos);
+  }, [todos, filter]);
+
+  const handleAddTodo = useCallback(async (todoTitle: string) => {
+    try {
+      const newTodo = {
+        userId: USER_ID,
+        title: todoTitle,
+        completed: false,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(currentTodos => [...currentTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  }, []);
+
+  const handleRemoveTodo = useCallback(async (todoId: number) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoId]);
+      await deleteTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setTempTodo(null);
+      setLoadingTodos([0]);
+    }
+  }, []);
+
+  const handleToggleTodoStatus = useCallback(async (todoToUpdate: Todo) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoToUpdate.id]);
+
+      const newTodoStatus = !todoToUpdate.completed;
+      const updatedTodo = await updateTodoStatus(
+        todoToUpdate.id,
+        newTodoStatus,
+      ) as Todo;
+
+      setTodos(prevTodos => (
+        prevTodos.map(todo => (
+          todo.id !== todoToUpdate.id
+            ? todo
+            : updatedTodo
+        ))
+      ));
+    } catch {
+      setError('Unable to update todo');
+    } finally {
+      setLoadingTodos(currentTodosId => currentTodosId
+        .filter(todoId => todoId !== todoToUpdate.id));
+    }
+  }, []);
+
+  const handleUpdateTodoTitle = useCallback(async (
+    todoToUpdate: Todo,
+    title: string,
+  ) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoToUpdate.id]);
+
+      const hasTitle = title
+        ? { title }
+        : { completed: !todoToUpdate.completed };
+
+      const updatedTodo = {
+        ...todoToUpdate,
+        ...hasTitle,
+      } as Todo;
+
+      await updateTodoTitle(updatedTodo);
+
+      setTodos(prevTodos => (
+        prevTodos.map(todo => (
+          todo.id !== todoToUpdate.id
+            ? todo
+            : updatedTodo
+        ))
```

Comment:
A similar piece of code is on lines 113-119, so you could put this logic in a separate handler and reuse it.